### PR TITLE
Support nodestyle callbacks

### PR DESCRIPTION
**Pull request by vaishnav98**
support both styles of callback based on the callback arguments length (as suggested by @mvduin)
https://github.com/beagleboard/bonescript/issues/38

test script based on bone101 examples : https://gist.github.com/vaishnav98/bda1c2e0355e584830e8c6daf8730f93

### CODE REVIEW

**Comment by mvduin** [comment]
Note that I'm still of the opinion that for pinmux and gpio functions the use of callbacks should be avoided and explicitly deprecated, hence I don't endorse further complicating the code by supporting new-style callbacks for those functions.

Only functions which can actually block should support callbacks.

**Comment by jadonk** [comment]
All functions need callbacks to be supported via the RPC mechanism. The math functions are expected to be done locally, so they don't need callbacks. Ideally, bonescript.js would be updated to include the math functions locally.

I do support sending out deprecation messages and updating the bone101 demo code to use new callback styles that are consistent... I just want to continue to test against the old code for a while and work with the old-style for a while to give folks a chance to migrate.

File: src/hw_mainline.js [hunk 1]
```diff
@@ -62,17 +62,26 @@ var readPinMux = function (pin, mode, callback) {
         if (err) {
             mode.err = 'readPinctrl error: ' + err;
             if (debug) winston.debug(mode.err);
-            callback(mode);
+            if (callback.length == 1)
```
Comment (jadonk) [inline src/hw_mainline.js]: When called via RPC, is the length transferred properly? I don't remember the exact process, but want to make sure this works in both the local and RPC cases.

File: src/hw_mainline.js [hunk 2]
```diff
@@ -62,17 +62,26 @@ var readPinMux = function (pin, mode, callback) {
...
     var tryPinctrl = function (exists) {
         if (exists) {
             fs.readFile(pinctrlFile, 'utf8', readPinctrl);
         } else {
             if (debug) winston.debug('getPinMode(' + pin.key + '): no valid mux data');
-            callback(mode);
+            if (callback.length == 1)
+                callback(mode);
+            else
+                callback('getPinMode(' + pin.key + '): no valid mux data', mode);
```
Comment (jadonk) [inline src/hw_mainline.js]: Can we try to get some consistency in the error messages? Can you check if all the others are in this same format?

**Comment by mvduin** [comment]
Ew, bonescript is also used via rpc from the browser instead of just locally as a nodejs library? That sounds... ehm... fast, efficient, and secure? ;-)

**Comment by jadonk** [comment]
Before I merge, please:
* Add deprecation notices to all old-style callbacks via 'warn' level winston calls.
* Add an RPC test bench (starts the server, then runs some basic tests) to get the coveralls test coverage high enough that coverage increases.

File: src/hw_mainline.js (hunk 1 above)
Comment (vaishnav98) [reply src/hw_mainline.js]: I did not test it previously via RPC , now when i tested : the nodestyle callback does not work properly via RPC , the previous style of callback work without any issues, will look more into this.

**Comment by vaishnav98** [comment]
I will work on the suggestions and make the neccessary changes soon

File: src/hw_mainline.js (hunk 1 above)
Comment (mvduin) [reply src/hw_mainline.js]: Both the [RPC client](https://github.com/jadonk/bonescript/blob/451348bc/src/bonescript.js#L73-L77) and [RPC server](https://github.com/jadonk/bonescript/blob/451348bc/src/socket_handlers.js#L60-L91) make hardcoded use of the old callback style. The client will need to be fixed to accept both types of callback handler and do the necessary translation. If the old callback style is to be deprecated, the server will need to be fixed to use node-style callbacks.

BTW, while taking a quick peek at this RPC code, I spotted an obvious memory leak:
```
dev:~/beagle/bonescript$ grep -r oneshot src/
src/socket_handlers.js:                // TODO: consider setting 'oneshot'
src/bonescript.js:    if (data.oneshot) delete _bonescript._callbacks[data.seq];
```
In other words, callback handlers are never removed from the client's bookkeeping and the _bonescript._callbacks array will keep growing forever.

**Comment by jadonk** [comment]
Here's some code to test the RPC server. This should also provide some test framework for adding the verification.

https://github.com/jadonk/bonescript/commit/47454f7046ce573970789193d1b5e5561b33673f

**Comment by vaishnav98** [comment]
I have made the recommended changes, can you please review it.

File: test/test-rpc.js [hunk 3]
```diff
@@ -2,24 +2,118 @@ var server = require('../src/server');
 var bonescript = require('../src/bonescript');
 var serverEmitter = null;
 
-exports.setUp = function (callback) {
+exports.testRPC = function (test) {
```
[the remark this answers is not in the document]
Comment (vaishnav98) [reply test/test-rpc.js]: I was assuming that all the rpc related tests should come under test-rpc.js , that's why i put all the test cases to a single module (as there could be more rpc tests in future), also i have a doubt regarding making the setup generic and have individual test case modules :  wouldn't the setUp be invoked for each test-case module, i.e try to create a server on the same port again which will throw : EADDRINUSE error unless the server is stopped per test-case through a suitable tearDown function ?
